openvdb_points/unittest/TestIndexFilter.cc

```diff
@@ -0,0 +1,812 @@
+template<class GridType>
+inline void
+makeSphere(const openvdb::Coord& dim, const openvdb::Vec3f& center, float radius, GridType& grid)
+{
+    typedef typename GridType::ValueType ValueT;
+    const ValueT zero = openvdb::zeroVal<ValueT>();
+
+    typename GridType::Accessor acc = grid.getAccessor();
+    openvdb::Coord xyz;
+    for (xyz[0]=0; xyz[0]<dim[0]; ++xyz[0]) {
+        for (xyz[1]=0; xyz[1]<dim[1]; ++xyz[1]) {
+            for (xyz[2]=0; xyz[2]<dim[2]; ++xyz[2]) {
+                const openvdb::Vec3R p =  grid.transform().indexToWorld(xyz);
+                const float dist = float((p-center).length() - radius);
+                ValueT val = ValueT(zero + dist);
+                acc.setValue(xyz, val);
+            }
+        }
+    }
+}
+
+
+void
+TestIndexFilter::testMultiGroupFilter()
+{
+    using namespace openvdb;
+    using namespace openvdb::tools;
+
+    typedef FilterIndexIter<IndexIter, MultiGroupFilter> IndexGroupIter;
+    typedef PointDataTree::LeafNodeType LeafNode;
+    typedef openvdb::tools::TypedAttributeArray<float>    AttributeS;
+
+    AttributeS::registerType();
+    GroupAttributeArray::registerType();
+
+    PointDataTree tree;
+    LeafNode* leaf = tree.touchLeaf(openvdb::Coord(0, 0, 0));
+
+    AttributeSet::Descriptor::Inserter names;
+    names.add("density", AttributeS::attributeType());
+    AttributeSet::Descriptor::Ptr descriptor = AttributeSet::Descriptor::create(names.vec);
+
+    leaf->initializeAttributes(descriptor, /*size=*/5);
+
+    appendGroup(tree, "even");
+    appendGroup(tree, "odd");
+    appendGroup(tree, "all");
+    appendGroup(tree, "first");
+
+    { // even
+        GroupWriteHandle groupHandle = leaf->groupWriteHandle("even");
+        groupHandle.set(0, true);
+        groupHandle.set(2, true);
+        groupHandle.set(4, true);
+    }
+
+    { // odd
+        GroupWriteHandle groupHandle = leaf->groupWriteHandle("odd");
+        groupHandle.set(1, true);
+        groupHandle.set(3, true);
+    }
+
+    setGroup(tree, "all", true);
+
+    { // first
+        GroupWriteHandle groupHandle = leaf->groupWriteHandle("first");
+        groupHandle.set(0, true);
+    }
+
+    IndexIter indexIter(0, 5);
+
+    { // all include
+        std::vector<Name> include; include.push_back("all");
+        std::vector<Name> exclude;
+        MultiGroupFilter::Data data(include, exclude);
+        MultiGroupFilter filter = MultiGroupFilter::create(*leaf, data);
+        IndexGroupIter iter(indexIter, filter);
+        CPPUNIT_ASSERT_EQUAL(iterCount(iter), size_t(5));
+    }
+
+    { // all exclude
+        std::vector<Name> include;
+        std::vector<Name> exclude; exclude.push_back("all");
+        MultiGroupFilter::Data data(include, exclude);
+        MultiGroupFilter filter = MultiGroupFilter::create(*leaf, data);
+        IndexGroupIter iter(indexIter, filter);
+        CPPUNIT_ASSERT_EQUAL(iterCount(iter), size_t(0));
+    }
```

Comment:
Would be good to add a test here to ensure the MultiGroupFilter works when both include and exclude are non-empty. For instance, including even but excluding first...